codec-http2/src/main/java/io/netty/handler/codec/http2/WeightedFairQueueByteDistributor.java

```diff
@@ -0,0 +1,297 @@
+import static io.netty.handler.codec.http2.Http2Error.INTERNAL_ERROR;
+import static io.netty.handler.codec.http2.Http2Exception.connectionError;
+import static io.netty.util.internal.ObjectUtil.checkNotNull;
+import static java.lang.Math.min;
+
+/**
+ * A {@link StreamByteDistributor} that is sensitive to stream priority and uses
+ * <a href="https://en.wikipedia.org/wiki/Weighted_fair_queueing">Weighted Fair Queueing</a> approach for distributing
+ * bytes.
+ * <p>
+ * Inspiration for this distributor was taken from Linux's
+ * <a href="https://git.kernel.org/cgit/linux/kernel/git/stable/linux-stable.git/tree/Documentation/scheduler
+ * /sched-design-CFS.txt">Completely Fair Scheduler</a>
+ * to model the distribution of bytes to simulate an "ideal multi-tasking CPU", but in this case we are simulating
+ * an "ideal multi-tasking NIC".
+ * <p>
+ * Each write operation will use the {@link #allocationQuantum(int)} to know how many more bytes should be allocated
+ * relative to the next stream which wants to write. This is to balance fairness while also considering goodput.
+ */
+public final class WeightedFairQueueByteDistributor implements StreamByteDistributor {
+    private final Http2Connection.PropertyKey stateKey;
+    private final State connectionState;
+    /**
+     * The minimum number of bytes that we will attempt to allocate to a stream. This is to
+     * help improve goodput on a per-stream basis.
+     */
+    private int allocationQuantum = 1024;
+
+    public WeightedFairQueueByteDistributor(Http2Connection connection) {
+        stateKey = connection.newKey();
+        Http2Stream connectionStream = connection.connectionStream();
+        connectionStream.setProperty(stateKey, connectionState = new State(connectionStream, 16, 0));
+
+        // Register for notification of new streams.
+        connection.addListener(new Http2ConnectionAdapter() {
+            @Override
+            public void onStreamAdded(Http2Stream stream) {
+                stream.setProperty(stateKey, new State(stream));
+            }
+
+            @Override
+            public void onWeightChanged(Http2Stream stream, short oldWeight) {
+                Http2Stream parent;
+                if (state(stream).isActiveCountForTree != 0 && (parent = stream.parent()) != null) {
+                    state(parent).totalQueuedWeights += stream.weight() - oldWeight;
+                }
+            }
+
+            @Override
+            public void onStreamClosed(Http2Stream stream) {
+                state(stream).close();
+            }
+
+            @Override
+            public void onPriorityTreeParentChanged(Http2Stream stream, Http2Stream oldParent) {
+                Http2Stream parent = stream.parent();
+                if (parent != null) {
+                    State state = state(stream);
+                    if (state.isActiveCountForTree != 0) {
+                        State pState = state(parent);
+                        pState.offer(state);
```

Comment:
There will need to be an initialization of `psuedoTime` equal to the parent's current time just before this queuing.